Step10 회원가입 화면 내 텍스트 검증 부분 구현

terry-yes opened the pull request:
## 회원가입 화면 내 텍스트 검증 부분 구현
> (미구현)4개의 텍스트필드의 조건이 다 성립하는 경우에만 다음버튼이 활성화되도록 구현 필요(현재는 조건이 안 맞아도 버튼이 눌림)
(미구현)필드에 입력된 내용을 바탕으로 네트워크 통신 보내는 부분 구현 필요

- 회원가입 화면의 텍스트필드와 라벨들을 재사용 가능하게 커스텀뷰로 만들어보았습니다.(CustomTextView.swift)
- 각 텍스트필드의 동작을 감지하는 textFieldDelegate들을 만들었습니다. 공통된 기능은 부모클래스로 만들고 각기 다르게 동작되는 부분은 상속한 자식클래스에서 구현하였습니다.
- 텍스트필드델리게이트에서 감지된 변화를 화면에 반영할 수 있도록 CustomTextView를 델리게이트화시켜서 갖고 있게 하였습니다.
  - 커스텀 텍스트 뷰의 텍스트필드가 텍스트필드델리게이트를 갖고 있고, 텍스트필드델리게이트가 프로퍼티로 텍스트필드를 갖고 있습니다.
  - 텍스트필드델리게이트에서 텍스트필드이외의 정보에는 접근할 수 없어서(텍스트필드 밑에 caption메세지에 접근할수없어서) 이렇게 델리게이트를 주는방식으로 해결했는데요. 이렇게 해도 되는지 잘 모르겠습니다
  - 아래는 구조를 그림으로 표현해봤습니다!(이해를 돕기위해 올리긴 했는데, 막상 올리고 보니...🥲)
![Screenshot_20220121-010632_Samsung Notes](https://user-images.githubusercontent.com/55867479/150376498-df9154a7-d6c9-4654-8563-51df9470e779.jpg)

- 작동화면은 [여기](https://github.com/terry-yes/w2-signup-profile/tree/step10)에 있습니다.

terry-yes committed:
feat: 탭바뷰컨트롤러전에 회원가입 씬 생성

terry-yes committed:
feat: 회원가입 화면 View 구현

- Custom View를 활용하여 반복되는 textField 뷰 구현

terry-yes committed:
feat: 회원가입 화면의 각 textFieldView에 맞는 이름, 캡션, placeholder설정

terry-yes committed:
feat: 기존 ViewController들 폴더로 정리

terry-yes committed:
feat: 회원가입 화면의 각각의 textField들의 검증 기능 구현

terry-yes committed:
feat: 회원가입 화면의 비밀번호 입력 텍스트필드의 값이 안보이게 처리

terry-yes committed:
docs: 현재까지 구현 사항 README.md에 정리

terry-yes committed:
docs: 비밀번호 필드와 비밀번호 재확인 필드에 실제로는 숫자가 보이지 않는다는 내용 추가

kyoo-sl commented on KakaoProfile/KakaoProfile/View/CustomTextView.swift:
```diff
@@ -0,0 +1,62 @@
+    func updateTitle(title: String)
+    func updateCaption(caption: String, color: UIColor)
+    func updateTextField(color: CGColor)
+}
```
TextViewDelegate는 다른 곳에서도 사용하지 않는다면, protocol로 굳이 빼지 않고 CustomTextView의 오픈 메소드로 제공해도 될 것 같아요.

terry-yes replied:
그렇군요!! 해당 프로토콜은 삭제했습니다!😄

kyoo-sl commented on KakaoProfile/KakaoProfile/Controller/SignUp/SignUpViewController.swift:
```diff
@@ -0,0 +1,53 @@
+    @IBOutlet weak var idTextView: CustomTextView!
+    @IBOutlet weak var passwordTextView: CustomTextView!
+    @IBOutlet weak var passwordCheckTextView: CustomTextView!
+    @IBOutlet weak var nameTextView: CustomTextView!
```
아무것도 입력하지 않고 한번씩 누르면 아래처럼 나오는 게 맞나요?
아무것도 입력하지 않앗을 때는 caption이나 보더 색상을 다시 normal하게 해주는 건 어떤가요?
<img src="https://user-images.githubusercontent.com/97584466/150451834-125d0aa4-4c70-4195-98d6-62838e84ca76.png" width="33%">

terry-yes replied:
말씀해주신 방식이 더 자연스러운거 같아서 바꾸었습니다!😄
마지막에 이름은 아무것도 없어야 캡션에 에러가 출력되어야 하는데, 이부분도 일단은 노말 상태가 되도록 하였습니다.
이후에 다음 버튼 눌렀을때도 이름이 빈칸이면 그때 에러메세지가 뜨게 구현해보겠습니다!

kyoo-sl commented on KakaoProfile/KakaoProfile/Controller/SignUp/TextFieldDelegate.swift:
```diff
@@ -0,0 +1,74 @@
+    
+    fileprivate func validate(text: String) -> (Bool, String) {
+        return (true, "")
+    }
```
TextFieldDelegate의 subclass로 제공하고 싶은건 validation 같은데요.
분기되어 각각 처리되는 동작을 제공할 수 있는 형태가 subclass말고 다른 방법이 있을 것 같아요.
예를 들어 validation을 closure 프로퍼티로 가지고 있고, TextFieldDelegate를 생성할 때 동작을 지정해 놓을 수 있을 것 같아요.

subclass로 제공하니 class가 많아지고 관계도 복잡해지는 것 같아요.

terry-yes replied:
이 부분을 정말 많이 고민했었는데, 클로저는 생각을 못해봤습니다!!
클로저로 바꾸었더니 훨씬 깔끔해져서 너무 좋아요
또한 클로저로 바꾸니 VC가 갖고 있는 데이터를 전달하기 더욱 수월해진거 같습니다. 
텍스트필드델리게이트가 텍스트뷰를 갖고 있는 현재구조도 왠지 클로저로 데이터를 전달해주면 더 깔끔하게 변경할 수 있을것 같습니다!!
이후에 더 간결하게 바꿔보겠습니다.
리뷰 감사드려요 @kyoo-sl !😁

terry-yes committed:
fix: 텍스트필드에 아무것도 입력을 안하고 다른 필드로 넘어갔을 때 borderCorlor나 caption이 normal상태로 돌아가게 구현

terry-yes committed:
refactor: 코드리뷰 사항 반영

- 한곳에서만 사용하는 델리게이트 프로토콜을 삭제
- 공통된 부분이 있는 텍스트필드델리게이트를 기존에는 상속으로 구현했는데 현재는 클로저를 전달하는 방식으로 변경

kyoo-sl reviewed (commented):
헛? 혹시 제가 approve를 안해서 아직 안 닫혓나요 ㅜ
제가 목/금에 리뷰를 해서, 혹시 다음 단계가 있다면 머지해도 좋을 것 같아요!
다음 단계가 있다면 다음 pr에서 추가로 볼 수 있습니다

terry-yes commented:
> 헛? 혹시 제가 approve를 안해서 아직 안 닫혓나요 ㅜ 제가 목/금에 리뷰를 해서, 혹시 다음 단계가 있다면 머지해도 좋을 것 같아요! 다음 단계가 있다면 다음 pr에서 추가로 볼 수 있습니다

앗 아니에요!
이 과제는 여기까지만 구현하고 week03 과제 하고 있었습니다!
